text/profiles/0237-profiles-data-model.md

```diff
@@ -0,0 +1,1010 @@
+  // all zeroes is considered invalid. Profile ID can be used by other signals to uniquely identify a profile.
+  //
+  // This field is required.
+  bytes profile_id = 1;
+
+  // start_time_unix_nano is the start time of the profile.
+  // Value is UNIX Epoch time in nanoseconds since 00:00:00 UTC on 1 January 1970.
+  //
+  // This field is semantically required and it is expected that end_time >= start_time.
+  fixed64 start_time_unix_nano = 2;
+
+  // end_time_unix_nano is the end time of the profile.
+  // Value is UNIX Epoch time in nanoseconds since 00:00:00 UTC on 1 January 1970.
+  //
+  // This field is semantically required and it is expected that end_time >= start_time.
+  fixed64 end_time_unix_nano = 3;
+
+  // attributes is a collection of key/value pairs. Note, global attributes
+  // like server name can be set using the resource API. Examples of attributes:
+  //
+  //     "/http/user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
+  //     "/http/server_latency": 300
+  //     "abc.com/myattribute": true
+  //     "abc.com/score": 10.239
+  //
+  // The OpenTelemetry API specification further restricts the allowed value types:
+  // https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute
+  // Attribute keys MUST be unique (it is not allowed to have more than one
+  // attribute with the same key).
+  repeated opentelemetry.proto.common.v1.KeyValue attributes = 4;
+
+  // dropped_attributes_count is the number of attributes that were discarded. Attributes
+  // can be discarded because their keys are too long or because there are too many
+  // attributes. If this value is 0, then no attributes were dropped.
+  uint32 dropped_attributes_count = 5;
+
+  // This is the original profile as retrieved from the profiler. For example, this can be a pprof or jfr encoded profile. The reason users might want to include these is because some formats are very generic and can not be easily converted to a more structured format.
+  // TODO: add a field that indicates the format of the original payload?
+  bytes original_payload = 6;
+
+  // A lookup table of Stacktraces. Other messages refer to Stacktraces in this table by index.
+  repeated Stacktrace stacktraces = 7;
+
+  // A lookup table of Mappings. Other messages refer to Mappings in this table by index.
+  repeated Mapping mappings = 8;
+
+  // A lookup table of Locations. Other messages refer to Locations in this table by index.
+  repeated Location locations = 9;
+
+  // A lookup table of Functions. Other messages refer to Functions in this table by index.
+  repeated Function functions = 10;
+
+  // A lookup table of Links to trace spans associated with this profile. Other messages refer to Links in this table by index. The first message must be an empty Link — this represents a null Link.
+  repeated Link links = 11;
```

Comment:
Could you clarify what you mean by colinear?

If you mean something like "2 arrays where each element with index `i` in the first array corresponds to an element with the same index `i` in the second array" then no, none of these are like that — all of these arrays are independent.

I think in practice `links` and `attribute_sets` could be colinear for some profiles, e.g I imagine it would be common that for a unique attribute set, e.g `request_id=123`, there would be a unique matching link, e.g `trace_id=xyz, span_id=abc`. But this is a bit a of stretch, and it's definitely not guaranteed that links and attribute sets would be colinear.

Arrays in `ProfileType` message are colinear.

Reply:
Sounds good thanks. (and sorry for confusing wording, I guess I should have said "parallel arrays").